File: core/node/paths.js
'use strict';

const path = require('path');

/**
 * Where a run reads and writes. The single place the core learns about the
 * filesystem, so the same package can serve every domain without knowing any
 * of them.
 *
 * The core is installed once and required from several places. Resolving
 * against __dirname would point every domain at the core's own folder; the
 * queue of one domain would then overwrite the queue of another. So the root
 * is the WORKING DIRECTORY of the process -- the domain's node/ folder, which
 * is where every npm script runs -- with an override for callers that run from
 * somewhere else (CI steps, the pipeline, a test).
 *
 *   workdir      the domain's node/ folder      queue/  reports/  state/
 *   domainRoot   one level up                   fixtures/  features/
 */

function workdir() {
  return path.resolve(process.env.QA_WORKDIR || process.cwd());
}

function domainRoot() {
  return path.resolve(process.env.QA_DOMAIN_ROOT || path.join(workdir(), '..'));
}

function queueFile() { return path.join(workdir(), 'queue', 'results.jsonl'); }
function reportsDir() { return path.join(workdir(), 'reports'); }
function stateDir() { return path.join(workdir(), 'state'); }
function issuesFile() { return path.join(stateDir(), 'issues.json'); }
function summaryFile() { return path.join(reportsDir(), 'summary.json'); }
function dashboardFile() { return path.join(reportsDir(), 'dashboard.html'); }

function fixturesDir() { return path.join(domainRoot(), 'fixtures'); }
function expectedFile() { return path.join(fixturesDir(), 'expected-results.json'); }
function testcasesFile() { return path.join(fixturesDir(), 'testcases.json'); }

module.exports = {
  workdir, domainRoot,
  queueFile, reportsDir, stateDir, issuesFile, summaryFile, dashboardFile,
  fixturesDir, expectedFile, testcasesFile,
};

File: core/python/qa_core/paths.py
"""Where a run reads and writes -- the one place the core learns about the filesystem.

The core is installed once (``pip install -e``) and imported from several
domains. Resolving against ``__file__`` would point every domain at the core's
own folder, and the queue of one domain would overwrite the queue of another.
So the root is the WORKING DIRECTORY of the process -- the domain's ``python/``
folder, where pytest runs -- with an override for callers that run elsewhere.

    workdir      the domain's python/ folder     queue/  reports/  state/
    domain_root  one level up                    fixtures/  features/
"""

from __future__ import annotations

import os
from pathlib import Path


def workdir() -> Path:
    return Path(os.environ.get("QA_WORKDIR") or os.getcwd()).resolve()


def domain_root() -> Path:
    override = os.environ.get("QA_DOMAIN_ROOT")
    return Path(override).resolve() if override else workdir().parent


def queue_file() -> Path:
    return workdir() / "queue" / "results.jsonl"


def reports_dir() -> Path:
    return workdir() / "reports"


def state_dir() -> Path:
    return workdir() / "state"


def issues_file() -> Path:
    return state_dir() / "issues.json"


def summary_file() -> Path:
    return reports_dir() / "summary.json"


def dashboard_file() -> Path:
    return reports_dir() / "dashboard.html"


def fixtures_dir() -> Path:
    return domain_root() / "fixtures"


def expected_file() -> Path:
    return fixtures_dir() / "expected-results.json"


def testcases_file() -> Path:
    return fixtures_dir() / "testcases.json"
